File: src/main/java/de/flavormate/ba_entities/tag/repository/TagRepository.java
/* Licensed under AGPLv3 2024 */
package de.flavormate.ba_entities.tag.repository;

import de.flavormate.ba_entities.recipe.model.Recipe;
import de.flavormate.ba_entities.tag.model.Tag;
import java.util.List;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface TagRepository extends JpaRepository<Tag, Long> {

  Optional<Tag> findByLabel(String label);

  @Query("select t from Tag t where t.recipes is empty")
  List<Tag> findAllEmpty();

  List<Tag> findAllByLabelIn(List<String> list);

  @Query(
      value = "select t from Tag t where lower(t.label) like lower(concat('%', :searchTerm, '%'))")
  Page<Tag> findBySearch(@Param("searchTerm") String searchTerm, Pageable pageable);

  @Query("select t.recipes from Tag t where t.id = ?1")
  Page<Recipe> findRecipesFromParent(Long id, Pageable pageable);
}

File: src/main/kotlin/de/flavormate/features/tag/repositories/TagRepository.kt
/* Licensed under AGPLv3 2024 - 2026 */
package de.flavormate.features.tag.repositories

import de.flavormate.features.recipe.daos.models.RecipeEntity
import de.flavormate.features.tag.daos.models.TagEntity
import io.quarkus.hibernate.orm.panache.kotlin.PanacheQuery
import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepositoryBase
import io.quarkus.panache.common.Sort
import jakarta.enterprise.context.ApplicationScoped

@ApplicationScoped
class TagRepository : PanacheRepositoryBase<TagEntity, String> {
  override fun findAll(sort: Sort): PanacheQuery<TagEntity> {
    return find("select t from TagEntity t", sort)
  }

  fun findByLabel(label: String): TagEntity? {
    return find("label", label).firstResult()
  }

  fun findBySearch(sort: Sort, query: String): PanacheQuery<TagEntity> {
    val params = mapOf("query" to query)

    return find(
      query = "select t from TagEntity t where lower(t.label) like lower(concat('%', :query, '%'))",
      sort = sort,
      params = params,
    )
  }

  fun findRecipesById(id: String, sort: Sort): PanacheQuery<RecipeEntity> {
    val params = mapOf("id" to id)
    return find(
        "select r from TagEntity t join t.recipes r where t.id = :id",
        sort = sort,
        params = params,
      )
      .project(RecipeEntity::class.java)
  }

  fun deleteOrphanTags(): Long {
    return delete("from TagEntity t where t.recipes is empty")
  }
}
